### PR TITLE
GH-4155 - Increase gRPC message size

### PR DESCRIPTION
**Pull request by komalali**
Fixes: #4155 

Related: https://github.com/pulumi/pulumi-policy/pull/239

Comments:
* I tried to minimize the number of individual places where that `1024 * 1024 * 400` number is referenced by creating a util function, but that only helps for the Go code. Is there a preferred way of adding such a widespread setting?
* Is the dotnet code adequate? Haven't ever worked in that lang before.
* For #4155 to truly be fixed, providers will need to be regenerated with pulumi-terraform-bridge that pulls this updated code and the related pulumi-policy PR must be merged.

### CODE REVIEW

File: sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs [hunk 1]
```diff
@@ -12,7 +13,9 @@ internal class GrpcEngine : IEngine
 
         public GrpcEngine(string engine)
         {
-            this._engine = new Engine.EngineClient(new Channel(engine, ChannelCredentials.Insecure));
+            var maxRpcMessageSize = 400 * 1024 * 1024;
```
Comment (lukehoban) [inline sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs:16]: Perhaps add a comment explaining this constant?  Here and in all the other similar locations.

File: sdk/go/pulumi/context.go [hunk 2]
```diff
@@ -27,13 +27,13 @@ import (
...
 	multierror "github.com/hashicorp/go-multierror"
-	"google.golang.org/grpc"
-
 	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
 	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/rpcutil"
 	pulumirpc "github.com/pulumi/pulumi/sdk/v2/proto/go"
+	"google.golang.org/grpc"
```
Comment (lukehoban) [inline sdk/go/pulumi/context.go:36]: Very minor - but we've generally maintained three blocks of imports (1) builtins (2) external dependencies (3) internal dependencies.

File: sdk/nodejs/runtime/settings.ts [hunk 3]
```diff
@@ -25,6 +26,9 @@ const resrpc = require("../proto/resource_grpc_pb.js");
 const resproto = require("../proto/resource_pb.js");
 const structproto = require("google-protobuf/google/protobuf/struct_pb.js");
 
+const maxRPCMessageSize: number = 1024 * 1024 * 400;
+const grpcChannelOptions = <ChannelOptions>{"grpc.max_receive_message_length": maxRPCMessageSize};
```
Comment (lukehoban) [inline sdk/nodejs/runtime/settings.ts:30]: Is the cast needed?  Or is this an option?

```suggestion
const grpcChannelOptions: ChannelOptions = {"grpc.max_receive_message_length": maxRPCMessageSize};
```

File: sdk/go/pulumi/context.go (hunk 2 above)
Comment (komalali) [reply sdk/go/pulumi/context.go:36]: That's what those 3 blocks represent! Okay good to know. Will fix.

File: sdk/nodejs/runtime/settings.ts (hunk 3 above)
Comment (komalali) [reply sdk/nodejs/runtime/settings.ts:30]: ah yup totally, typescript n00b moment.

File: sdk/nodejs/tests/runtime/langhost/run.spec.ts [hunk 4]
```diff
@@ -1136,6 +1136,24 @@ describe("rpc", () => {
...
+                assert.deepEqual(res, { "largeStringProp": longString });
+                return {
+                    urn: makeUrn(t, name),
+                    id: name,
+                    props: {
+                        "largeStringProp": "a".repeat(1024 * 1024 * 5),
+                    },
+                };
+            },
+        },
```
Comment (lukehoban) [inline sdk/nodejs/tests/runtime/langhost/run.spec.ts:1156]: Do these tests actually go through the RPC layer?  I don't recall off the top of my head - but I feel like they do not?  cc @pgavlin 

One place we could add tests to be sure they hit the RPC is in `/test/integration`.  Those will run the full "normal" CLI/Engine/LangHost setup.

Comment (lukehoban) [reply sdk/nodejs/tests/runtime/langhost/run.spec.ts:1156]: BTW - same likely applies to other languages.

Comment (komalali) [reply sdk/nodejs/tests/runtime/langhost/run.spec.ts:1156]: @lukehoban yes they do. I made sure by checking that they fail on master with the same resource exhausted error. I can definitely add integration tests too but I am certain that these tests go through the langhost <-> engine RPC layer.

**Comment by komalali** [comment]
Okay! I added in the integration tests here, added the skipped Node.js test in Windows to the list in https://github.com/pulumi/pulumi/issues/3811 and I think we're good to go here.

**Pull request merged**